src/poprox_recommender/handler.py

```diff
@@ -1,62 +1,110 @@
+    "Tech ",
+    "Lifestyle",
+    "Religion",
+    "Climate",
+    "Education",
+    "Oddities",
+]
 
-    checkpoint = load_file(load_path)
-    return checkpoint, device_name
 
+def generate_recs(event, context):
+    logger.info(f"Received event: {event}")
+    body = base64.b64decode(event["body"]) if event["isBase64Encoded"] else event["body"]
+    logger.info(f"Decoded body: {body}")
+    req = RecommendationRequest.model_validate_json(body)
 
-MODEL, DEVICE = load_model()
-TOKEN_MAPPING = "distilbert-base-uncased"  # can be modified
+    algo_params = event.get("queryStringParameters", {})
 
+    if algo_params:
+        logger.info(f"Using parameters: {algo_params}")
+    else:
+        logger.info("Using default parameters")
 
-def generate_recs(event, context):
-    req_body = json.loads(event["body"])
-
-    todays_articles = [
-        Article.model_validate(attrs) for attrs in req_body["todays_articles"]
-    ]
-    past_articles = [
-        Article.model_validate(attrs) for attrs in req_body["past_articles"]
-    ]
-    click_data = [
-        ClickHistory.model_validate(attrs) for attrs in req_body["click_data"]
-    ]
-    num_recs = req_body["num_recs"]
 
+    logger.info(f"Selecting articles...")
     recommendations = select_articles(
-        todays_articles,
-        past_articles,
-        click_data,
-        MODEL,
-        DEVICE,
-        TOKEN_MAPPING,
-        num_recs,
+        req.todays_articles,
+        req.past_articles,
+        req.click_histories,
+        req.num_recs,
+        algo_params,
     )
 
-    body = {
-        "recommendations": {
-            str(account_id): [
-                json.loads(article.model_dump_json()) for article in articles
-            ]
-            for account_id, articles in recommendations.items()
-        },
-    }
+    logger.info(f"Constructing response...")
+    resp_body = RecommendationResponse.model_validate(
+        {"recommendations": recommendations}
+    )
 
-    response = {"statusCode": 200, "body": json.dumps(body)}
+    logger.info(f"Serializing response...")
+    response = {"statusCode": 200, "body": resp_body.model_dump_json()}
 
+    logger.info(f"Finished.")
     return response
+
+def extract_general_topic(article):
+    news_topics = [mention.entity.name for mention in article.mentions]
+    article_topics = set()
+    for topic in news_topics:
+        if topic in general_topics:
+            article_topics.add(topic)
+    return article_topics
+
+def classify_news_topic(model, tokenizer, general_topics, topic):
+    inputs = tokenizer.batch_encode_plus([topic] + general_topics, return_tensors='pt', pad_to_max_length=True)
+    input_ids = inputs['input_ids']
+    attention_mask = inputs['attention_mask']
+    output = model(input_ids, attention_mask=attention_mask)[0]
+    
+    sentence_rep = output[:1].mean(dim=1)
+    label_reps = output[1:].mean(dim=1)
+
+    # now find the labels with the highest cosine similarities to the sentence
```

Comment:
💡 This is encoding the labels alongside the topic to be classified and then looking for similarity? I think that's what I was missing before

Reply:
Yep! And I manually tested on some news topics and confirmed they look accurate for sub-topic classification.